homeassistant/components/ezviz/__init__.py

```diff
@@ -50,54 +64,73 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    if PLATFORMS_BY_TYPE[sensor_type]:
+
+        # Get user account token if not present.
+        if not entry.data.get(CONF_SESSION_ID):
+
+            try:
+                ezviz_client = await _get_ezviz_client_instance(hass, entry)
 
-        return True
+            except (InvalidURL, HTTPError, PyEzvizError) as error:
+                _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
+                raise ConfigEntryNotReady from error
 
-    try:
-        ezviz_client = await hass.async_add_executor_job(
-            _get_ezviz_client_instance, entry
+        if not ezviz_client:
+            # No Ezviz login session, call api login().
+
+            ezviz_client = EzvizClient(
+                token={
+                    CONF_SESSION_ID: entry.data.get(CONF_SESSION_ID),
+                    CONF_RFSESSION_ID: entry.data.get(CONF_RFSESSION_ID),
+                    "api_url": entry.data.get(CONF_URL),
+                },
+                timeout=entry.options.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+            )
+
+            try:
+                await hass.async_add_executor_job(ezviz_client.login)
+
+            except (EzvizAuthTokenExpired, EzvizAuthVerificationCode) as error:
+                raise ConfigEntryAuthFailed from error
+
+            except (InvalidURL, HTTPError, PyEzvizError) as error:
+                _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
+                raise ConfigEntryNotReady from error
+
+        coordinator = EzvizDataUpdateCoordinator(
+            hass, api=ezviz_client, api_timeout=entry.options[CONF_TIMEOUT]
         )
-    except (InvalidURL, HTTPError, PyEzvizError) as error:
-        _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
-        raise ConfigEntryNotReady from error
 
-    coordinator = EzvizDataUpdateCoordinator(
-        hass, api=ezviz_client, api_timeout=entry.options[CONF_TIMEOUT]
-    )
-    await coordinator.async_refresh()
+        hass.data[DOMAIN][entry.entry_id] = {DATA_COORDINATOR: coordinator}
 
-    if not coordinator.last_update_success:
-        raise ConfigEntryNotReady
+        await coordinator.async_config_entry_first_refresh()
 
-    undo_listener = entry.add_update_listener(_async_update_listener)
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        DATA_COORDINATOR: coordinator,
-        DATA_UNDO_UPDATE_LISTENER: undo_listener,
-    }
-    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    if sensor_type == ATTR_TYPE_CAMERA:
+        if hass.data.get(DOMAIN):
+            for item in hass.config_entries.async_entries(domain=DOMAIN):
+                if item.data.get(CONF_TYPE) == ATTR_TYPE_CLOUD:
+                    _LOGGER.info("Reload Ezviz main account with camera entry")
+                    await hass.config_entries.async_reload(item.entry_id)
```

Comment:
What is this code doing? Wouldn't that mean that every camera entry is going to reload the cloud entry ? why even have a camera entry ?

Reply:
At present I'm using the camera entry to basically just store rtsp credentials for the camera. Had to reload for it to take effect if a user adds credentials for a camera.

There must be a better way of doing this, let me think about this one for a bit.

Reply:
Hi @balloob,

The new text entity might work to store the camera RTSP password. None the less, I will improve on a follow up pull request to keep this one as small as possible. The current behavior works similar to how the integration does it currently.  (Just later on to add update listener)